### DISCUSSION TITLE
Mishard proxy do_search process too slow

### DISCUSSION THREAD

**Discussion by OsmanFarukO** [milvus-io/milvus#10210] [Q&A and General discussion]
Hello everyone, im running mishard milvus with 4 ro-server (v1.1.0) and 1 proxy GPU enabled in kubernetes. I set `gpu_search_threshold: 1000` in server_config.yaml for use CPU+GPU together. I got ~50 million entity in milvus. When i search over mishard proxy its sharding very well over 4 ro-server instance ~2ms and do_merge process take under 1ms. But i dont understand why do_search process take ~1 second ? Here is Jaeger query trace output. Can anyone explain this for me :) Thank you.

![Screenshot from 2021-10-19 14-33-25](https://user-images.githubusercontent.com/24430724/137901415-91e858f5-66ed-4be0-890c-cda5de925e6a.png)

**Comment by yhmo** [comment]
The "do_routing" phase is to hash segment ids to different nodes.
The "do_search" phase is to wait nodes load segments, and search in segments
The "do_merge" phase is to combine results from nodes into one final result.

The time cost of "do_search": 1. load segments  2. search in segments
A few questions:
- What is the dimension of your vectors?
- What the value of the index_file_size? (this is a parameter of create_collection() )
- What the value of cache_size? (this is a config item in the server_config.yaml)

**Reply by OsmanFarukO** [reply]
Hi, dimension of my vector is 512. index_file_size is 2048 and cache size for CPU 150gb (90gb used), for each GPU cache size is 16gb. Thank you.

**Reply by yhmo** [reply]
50M * 512 * 4 = 100GB. 4 ro, each node takes 25GB. 

- Did you create index? Index parameters?
- Search parameters? how many nq?

**Reply by OsmanFarukO** [reply]
Yes, exactly each node takes 25GB. I created IVFFLAT index with 16384 nlist parameter. At the search nq=2 and nprobe=50. I tried nprobe 10, 100 , 128.

**Reply by yhmo** [reply]
The gpu_search_threshold = 1000, means when nq > 1000 use GPU to compute, otherwise use CPU to compute.
Search nq=2, it will use CPU to compute, the performance will depend on the CPU cores and ability.

You can check the xxx.debug.log under /var/lib/milvus/logs of a read-only node. Search "ExecutionEngineImpl::Search float" in the xxx.debug.log. Each ro node have 12 ~ 13 segments, each segment can generate some "ExecutionEngineImpl::Search float" log. We can know how much time is cost during searching.

**Reply by OsmanFarukO** [reply]
Ok there is ro-server-0 debug.log file. 

[servers-0.txt](https://github.com/milvus-io/milvus/files/7380634/servers-0.txt)

**Reply by yhmo** [reply]
From the log, I saw this ro node takes 16 segments(14 x 2Gb + 2 x 0.3GB). Each segment cost 40ms ~ 80ms to search, each request(nq=1, topk=2) cost 800ms ~ 1000ms. All segments use CPU to compute.
The performance of cpu computing is a bit slow. I guess the CPU cores is few.
